cardano-git-rev/src/Cardano/Git/Rev.hs

```diff
@@ -32,10 +32,11 @@ foreign import ccall "&_cardano_git_rev" c_gitrev :: CString
 -- This must be a TH splice to ensure the git commit is captured at build time.
 -- ie called as `$(gitRev)`.
 gitRev :: Q Exp
-gitRev
-  | gitRevEmbed /= zeroRev = textE gitRevEmbed
-  | otherwise              =
-      textE =<< TH.runIO runGitRevParse
+gitRev =
+    [| if
+         | gitRevEmbed /= zeroRev -> gitRevEmbed
+         | otherwise              -> $(textE =<< TH.runIO runGitRevParse)
+    |]
 
```

Comment:
Not sure about this. It has two TH embedded inside TH. How does the staging work?

I wonder if the correct solution isn't something like:
```
gitRev :: Text
gitRev
  | gitRevEmbed /= zeroRev = textE gitRevEmbed
  | otherwise              = $(textE =<< TH.runIO runGitRevParse)
```

Reply:
> Not sure about this. It has two TH embedded inside TH. How does the staging work?

In the end, `gitrev` will just be a quoted expression, and when spliced at the top-level at usage site, everything will happen in "one go" (in particular, `TH.runIO runGitRevParse` will also be run when spliced, not earlier when eg `cardano-git-rev` gets built).

---

Your snippet would also correctly run the `gitRevEmbed /= zeroRev` check at runtime, but the `$(textE =<< TH.runIO runGitRevParse)` spliced at the top-level, so it (and its embedded `IO`) will be run when building `cardano-git-rev`, which, when another project depends on `cardano-git-rev`, will be cached across multiple runs in the cabal store. This is in contrast to what happens in this PR, where we splice into another quotation, not at the top-level.

Before #467, the same problem was present, as `gitRevFromGit` was spliced at the top-level: 
https://github.com/IntersectMBO/cardano-base/blob/113ff707b91e059d7f41ec18b7c1ade481ab7bba/cardano-git-rev/src/Cardano/Git/Rev.hs#L48
Hence, the splice and the `IO` contained therein was executed while building `cardano-git-rev`.

Reply:
You have thoroughly convinced me I still have no idea how TH staging works.

Reply:
Maybe as another intuition: Here, in `cardano-git-rev`, we just want to provide expressions that users of this libraries can splice in their projects, but we don't actually need to splice anything at the top-level here ourselves. There actually is a subset of `TemplateHaskell` for that, [`TemplateHaskellQuotes`](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/template_haskell.html#extension-TemplateHaskellQuotes):

> Template Haskell has the following new syntactic constructions. You need to use the extension [`TemplateHaskell`](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/template_haskell.html#extension-TemplateHaskell) to switch these syntactic extensions on. Alternatively, the [`TemplateHaskellQuotes`](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/template_haskell.html#extension-TemplateHaskellQuotes) extension can be used to enable the quotation subset of Template Haskell (i.e. without top-level splices). The [`TemplateHaskellQuotes`](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/template_haskell.html#extension-TemplateHaskellQuotes) extension is considered safe under [Safe Haskell](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/safe_haskell.html#safe-haskell) while [`TemplateHaskell`](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/template_haskell.html#extension-TemplateHaskell) is not.

I added another commit that only enables `TemplateHaskellQuotes` here to emphasize that we don't (currently) want any top-level splices in `cardano-git-rev`.